File: 3gu4a1-Todo_list_app/evaluation/evaluation.py
#!/usr/bin/env python3
"""
Evaluation script for comparing repository_before and repository_after.
"""

import os
import sys
import json
import uuid
import re
import platform
import subprocess
from pathlib import Path
from datetime import datetime, timezone

ROOT = Path(__file__).resolve().parent.parent
REPORTS_BASE = ROOT / "evaluation" / "reports"


def environment_info():
    """Collect environment metadata."""
    return {
        "python_version": platform.python_version(),
        "platform": platform.platform(),
    }


def run_tests(repo_name: str):
    """Run pytest tests for a specific repository. Returns test results dictionary."""
    # Check for implementation - repository_after has app/main.py, repository_before may have main.py
    repo_path = ROOT / repo_name
    has_implementation = False
    
    if repo_name == "repository_after":
        # repository_after has app/main.py
        has_implementation = (repo_path / "app" / "main.py").exists()
    else:
        # repository_before may have main.py at root or app/main.py
        has_implementation = (repo_path / "main.py").exists() or (repo_path / "app" / "main.py").exists()
    
    if not has_implementation:
        return {
            "passed": False,
            "return_code": -1,
            "output": f"Repository {repo_name} has no implementation (main.py or app/main.py not found)",
        }

    try:
        env = os.environ.copy()
        env["PYTHONPATH"] = f"{ROOT}:{env.get('PYTHONPATH', '')}".rstrip(":")
        # Set environment variable to filter which repository to test
        env["TEST_REPOSITORY"] = repo_name

        proc = subprocess.run(
            ["pytest", "tests", "-q", "--tb=short"],
            cwd=ROOT,
            env=env,
            capture_output=True,
            text=True,
            timeout=120,
        )

        output = (proc.stdout + proc.stderr)[:8000]
        # Check for passed tests - must have at least one passed test
        has_passed = "passed" in output.lower() and "0 passed" not in output.lower()
        # For repository_after, ensure no tests were skipped due to import errors
        # Check that we have actual test results, not just skips
        if repo_name == "repository_after":
            # Count passed tests - should have significant number of passed tests
            passed_match = re.search(r"(\d+)\s+passed", output)
            skipped_match = re.search(r"(\d+)\s+skipped", output)
            passed_count = int(passed_match.group(1)) if passed_match else 0
            skipped_count = int(skipped_match.group(1)) if skipped_match else 0
            # For repository_after, we expect many passed tests and minimal skips (only for repository_before)
            # If all tests are skipped, that's a failure
            if passed_count == 0 and skipped_count > 0:
                has_passed = False
            # Also check for import errors that would cause skips
            if "Skipping repository_after" in output or "module not found" in output.lower():
                has_passed = False

        return {
            "passed": proc.returncode == 0 and has_passed,
            "return_code": proc.returncode,
            "output": output,
        }
    except subprocess.TimeoutExpired:
        return {
            "passed": False,
            "return_code": -1,
            "output": "pytest timeout after 120 seconds",
        }
    except Exception as e:
        return {
            "passed": False,
            "return_code": -1,
            "output": f"Error running tests: {str(e)}",
        }


def run_metrics():
    """
    Optional metrics collection.
    Implement task-specific metrics here if needed.
    """
    # Placeholder for future metrics
    # Example metrics could include:
    # - Performance metrics (response times, throughput)
    # - Stability metrics (failure rates)
    # - Resource usage (memory, CPU)
    return {}


def evaluate(repo_name: str):
    """
    Evaluate a repository by running tests and collecting metrics.
    Returns evaluation results dictionary.
    """
    # Run tests
    tests = run_tests(repo_name)

    # Collect metrics (optional)
    metrics = run_metrics()

    return {"tests": tests, "metrics": metrics}


def run_evaluation():
    """
    Main evaluation function.
    Runs evaluation on both repositories and generates comparison report.
    Returns complete report dictionary.
    """
    run_id = str(uuid.uuid4())
    start = datetime.now(timezone.utc)

    try:
        # Evaluate repository_before
        before = evaluate("repository_before")

        # Evaluate repository_after
        after = evaluate("repository_after")

        # Compare results
        before_passed = before["tests"]["passed"]
        after_passed = after["tests"]["passed"]

        summaries = {
            (True, True): "Both implementations passed tests",
            (False, True): "After implementation fixed failing tests",
            (True, False): "After implementation introduced regressions",
            (False, False): "Both implementations failed tests",
        }

        comparison = {
            "passed_gate": after_passed,
            "improvement_summary": summaries[(before_passed, after_passed)],
        }

        end = datetime.now(timezone.utc)
        duration = (end - start).total_seconds()

        report = {
            "run_id": run_id,
            "started_at": start.isoformat().replace("+00:00", "Z"),
            "finished_at": end.isoformat().replace("+00:00", "Z"),
            "duration_seconds": duration,
            "environment": environment_info(),
            "before": before,
            "after": after,
            "comparison": comparison,
            "success": comparison["passed_gate"],
            "error": None,
        }

        return report

    except Exception as e:
        # Error handling: evaluation crashed
        end = datetime.now(timezone.utc)
        duration = (end - start).total_seconds()

        return {
            "run_id": run_id,
            "started_at": start.isoformat().replace("+00:00", "Z"),
            "finished_at": end.isoformat().replace("+00:00", "Z"),
            "duration_seconds": duration,
            "environment": environment_info(),
            "before": {
                "tests": {"passed": False, "return_code": -1, "output": ""},
                "metrics": {},
            },
            "after": {
                "tests": {"passed": False, "return_code": -1, "output": ""},
                "metrics": {},
            },
            "comparison": {
                "passed_gate": False,
                "improvement_summary": "Evaluation crashed",
            },
            "success": False,
            "error": str(e),
        }


def main():
    """
    Main entry point.
    Creates reports directory, runs evaluation, writes report, and returns exit code.
    """
    # Run evaluation
    report = run_evaluation()

    # Create nested directory structure: YYYY-MM-DD/HH-MM-SS/
    now = datetime.now(timezone.utc)
    date_dir = now.strftime("%Y-%m-%d")
    time_dir = now.strftime("%H-%M-%S")
    report_dir = REPORTS_BASE / date_dir / time_dir
    report_dir.mkdir(parents=True, exist_ok=True)

    # Write report to report.json
    report_path = report_dir / "report.json"
    report_path.write_text(json.dumps(report, indent=2))

    print(f"Evaluation report written to {report_path}")
    print(f"Success: {report['success']}")
    print(f"Before tests passed: {report['before']['tests']['passed']}")
    print(f"After tests passed: {report['after']['tests']['passed']}")

    # Return exit code: 0 for success, 1 for failure
    return 0 if report["success"] else 1


if __name__ == "__main__":
    sys.exit(main())

File: c8f2a3-Stackable_Promo_Codes_Feature/evaluation/evaluation.js
#!/usr/bin/env node
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const { randomUUID } = require('crypto');

const ROOT = path.resolve(__dirname, '..');
const REPORTS = path.join(ROOT, 'evaluation', 'reports');

function environmentInfo() {
  return {
    node_version: process.version,
    platform: process.platform,
    arch: process.arch
  };
}

function runTests(repoName) {
  const repoPath = path.join(ROOT, repoName);
  
  try {
    // Run jest from workspace root directory
    const output = execSync(
      `npx jest --runInBand --forceExit --json`,
      {
        cwd: ROOT,
        env: {
          ...process.env,
          REPO_PATH: `../${repoName}`,
          MONGODB_URI: process.env.MONGODB_URI || 'mongodb://localhost:27017',
          MONGODB_DB: process.env.MONGODB_DB || 'test_promocodes',
          NODE_PATH: repoPath
        },
        encoding: 'utf-8',
        timeout: 300000,
        stdio: 'pipe'
      }
    );
    
    try {
      const jestOutput = JSON.parse(output);
      const passed = jestOutput.success || false;
      const numPassed = jestOutput.numPassedTests || 0;
      const numFailed = jestOutput.numFailedTests || 0;
      const numTotal = jestOutput.numTotalTests || 0;
      
      return {
        passed: passed,
        return_code: passed ? 0 : 1,
        output: `Tests: ${numPassed}/${numTotal} passed${numFailed > 0 ? `, ${numFailed} failed` : ''}`
      };
    } catch (parseError) {
      // Fallback if JSON parsing fails
      return {
        passed: false,
        return_code: 1,
        output: output.substring(0, 8000)
      };
    }
  } catch (error) {
    const errorOutput = error.stdout || error.stderr || error.message || '';
    return {
      passed: false,
      return_code: error.status || -1,
      output: errorOutput.substring(0, 8000)
    };
  }
}

function runMetrics(repoPath) {
  // Optional metrics collection - not implemented for this task
  return {};
}

function evaluate(repoName) {
  const repoPath = path.join(ROOT, repoName);
  
  // Install dependencies first (only if package.json exists)
  const packageJsonPath = path.join(repoPath, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    try {
      execSync('npm install', {
        cwd: repoPath,
        encoding: 'utf-8',
        timeout: 120000,
        stdio: 'pipe'
      });
    } catch (error) {
      return {
        tests: {
          passed: false,
          return_code: -1,
          output: `npm install failed: ${(error.stderr || error.message || '').substring(0, 2000)}`
        },
        metrics: {}
      };
    }
  }
  
  // Run tests
  const tests = runTests(repoName);
  
  // Collect metrics (optional)
  const metrics = runMetrics(repoPath);
  
  return {
    tests: tests,
    metrics: metrics
  };
}

function runEvaluation() {
  const runId = randomUUID();
  const start = new Date();
  
  try {
    // Evaluate both repositories
    const before = evaluate('repository_before');
    const after = evaluate('repository_after');
    
    // Compare results
    const passedGate = after.tests.passed;
    
    const beforePassed = before.tests.passed || false;
    const afterPassed = after.tests.passed || false;
    
    let improvementSummary;
    if (afterPassed && !beforePassed) {
      improvementSummary = "After implementation passes all tests, before implementation fails (expected - method doesn't exist)";
    } else if (afterPassed && beforePassed) {
      improvementSummary = "Both implementations pass tests";
    } else if (!afterPassed && beforePassed) {
      improvementSummary = "Regression: after implementation fails tests";
    } else {
      improvementSummary = "Both implementations fail tests";
    }
    
    const comparison = {
      passed_gate: passedGate,
      improvement_summary: improvementSummary
    };
    
    const end = new Date();
    const duration = (end - start) / 1000;
    
    return {
      run_id: runId,
      started_at: start.toISOString(),
      finished_at: end.toISOString(),
      duration_seconds: duration,
      environment: environmentInfo(),
      before: before,
      after: after,
      comparison: comparison,
      success: comparison.passed_gate,
      error: null
    };
  } catch (error) {
    const end = new Date();
    return {
      run_id: runId,
      started_at: start.toISOString(),
      finished_at: end.toISOString(),
      duration_seconds: (end - start) / 1000,
      environment: environmentInfo(),
      before: {
        tests: { passed: false, return_code: -1, output: "" },
        metrics: {}
      },
      after: {
        tests: { passed: false, return_code: -1, output: "" },
        metrics: {}
      },
      comparison: {
        passed_gate: false,
        improvement_summary: "Evaluation error occurred"
      },
      success: false,
      error: error.message || String(error)
    };
  }
}

function main() {
  // Create reports directory structure
  if (!fs.existsSync(REPORTS)) {
    fs.mkdirSync(REPORTS, { recursive: true });
  }
  
  // Generate report
  const report = runEvaluation();
  
  // Create date/time directory structure
  const now = new Date();
  const dateDir = now.toISOString().split('T')[0]; // YYYY-MM-DD
  const timeDir = now.toTimeString().split(' ')[0].replace(/:/g, '-'); // HH-MM-SS
  const reportDir = path.join(REPORTS, dateDir, timeDir);
  
  if (!fs.existsSync(reportDir)) {
    fs.mkdirSync(reportDir, { recursive: true });
  }
  
  // Write report to date/time directory
  const reportPath = path.join(reportDir, 'report.json');
  fs.writeFileSync(reportPath, JSON.stringify(report, null, 2));
  
  console.log(`Report written to ${reportPath}`);
  console.log(`Success: ${report.success}`);
  console.log(`Before tests passed: ${report.before.tests.passed}`);
  console.log(`After tests passed: ${report.after.tests.passed}`);
  
  return report.success ? 0 : 1;
}

if (require.main === module) {
  try {
    process.exit(main());
  } catch (error) {
    console.error('Evaluation error:', error);
    process.exit(1);
  }
}

module.exports = { runEvaluation, main };
